src/star-id.cpp

```diff
@@ -1,21 +1,346 @@
+Start from index and does quadratic probing
+*/
+std::vector<std::vector<int>> TetraStarIdAlgorithm::GetAtIndex(int index, int maxIndex,
+                                                               const TetraDatabase &db) const {
+    std::vector<std::vector<int>> res;
+    for (int c = 0;; c++) {
+        int i = (index + c * c) % maxIndex;
+
+        std::vector<int> tableRow = db.GetPattern(i);
+
+        if (tableRow[0] == 0 && tableRow[1] == 0) {
+            break;
+        } else {
+            res.push_back(tableRow);
+        }
+    }
+    return res;
+}
+
+static bool GetCentroidCombination(int pattSize, int numCentroids, bool firstTime,
+                                   std::vector<int> &indices, std::vector<int> *const res) {
+    if (numCentroids < pattSize) {
+        return false;
+    }
+
+    if (firstTime) {
+        firstTime = false;
+        indices.push_back(-1);
+        for (int i = 0; i < pattSize; i++) {
+            indices.push_back(i);
+        }
+        indices.push_back(numCentroids);
+        *res = std::vector<int>(indices.begin() + 1, indices.end() - 1);
+        return true;
+    }
+
+    if (indices[1] >= numCentroids - pattSize) {
+        return false;
+    }
+
+    for (int i = 1; i <= pattSize; i++) {
+        indices[i]++;
+        if (indices[i] < indices[i + 1]) {
+            break;
+        }
+        indices[i] = indices[i - 1] + 1;
+    }
+    *res = std::vector<int>(indices.begin() + 1, indices.end() - 1);
+    return true;
+}
+
+/**
+ * @brief Identifies stars using Tetra Star Identification Algorithm
+ *
+ * @param database
+ * @param stars List of star centroids detected from image
+ * @param catalog Star table
+ * @param camera
+ * @return StarIdentifiers
+ */
+StarIdentifiers TetraStarIdAlgorithm::Go(const unsigned char *database, const Stars &stars,
+                                         const Catalog &catalog, const Camera &camera) const {
+    // Result format: (centroidIndex, catalogIndex)
+    StarIdentifiers result;
+    std::cout << "TETRA" << std::endl;
+
+    ////////////////////////////////////////////////////////
+
+    MultiDatabase multiDatabase(database);
+    const unsigned char *databaseBuffer =
+        multiDatabase.SubDatabasePointer(TetraDatabase::kMagicValue);
+    if (databaseBuffer == NULL) {
+        std::cerr << "Could not get to Tetra database" << std::endl;
+        return result;
+    }
+    TetraDatabase tetraDatabase(databaseBuffer);
+
+    const long long catLength = tetraDatabase.Size();
+    const float maxFov = tetraDatabase.MaxAngle();
+
+    // std::cout << "Tetra can handle max FOV: " << maxFov << std::endl;
+
+    std::vector<int> centroidIndices;
+    for (int i = 0; i < (int)stars.size(); i++) {
+        centroidIndices.push_back(i);
+    }
+
+    // Sort centroided stars by brightness, high to low. Larger is brighter
+
+    std::stable_sort(centroidIndices.begin(), centroidIndices.end(),
```

Comment:
why stable? (It's not a big deal for performance, it just confuses readers to use stable if you don't need it).

Reply:
Fixed